File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void preorderTraversal_01(TreeNode node,ArrayList<Integer> ans) {
        ans.add(node.val);
        
        if(node.left != null){
            preorderTraversal_01(node.left,ans);
        }
        
        if(node.right != null){
            preorderTraversal_01(node.right,ans);
         }
    }
    
    public List<Integer> preorderTraversal(TreeNode root) {
        //Base case
        ArrayList<Integer> list = new ArrayList<>();
        if(root==null)  return list;
        preorderTraversal_01(root,list);
        
        return list;
    }
}

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        if(root == nullptr) return {};
        vector<int> list;
        
        stack<TreeNode*> stack;
        stack.push(root);
        while(stack.size()!=0) {
            TreeNode* rn = stack.top(); stack.pop();
            
            if(rn->right != nullptr) {
                stack.push(rn->right);
            }
            
            if(rn->left != nullptr) {
                stack.push(rn->left);
            }
            
            list.push_back(rn->val);
        }
        
        return list;
    }
};